File: bridges/python/src/sdk/tools/ytdlp_tool.py
import subprocess
import os
from ..base_tool import BaseTool
from ..toolkit_config import ToolkitConfig


class YtdlpTool(BaseTool):
    TOOLKIT = 'video_streaming'

    def __init__(self):
        super().__init__()
        # Load configuration from central toolkits directory
        # Use class name for tool config name
        tool_config_name = self.__class__.__name__.lower().replace('tool', '')
        self.config = ToolkitConfig.load(self.TOOLKIT, tool_config_name)

    @property
    def tool_name(self) -> str:
        # Dynamic tool name based on class name
        return self.__class__.__name__

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config['description']

    def download_video(self, video_url: str, output_path: str) -> str:
        """
        Downloads a single video from the provided URL.
        
        Args:
            video_url: The URL of the video to download.
            output_path: The directory where the video will be saved.
            
        Returns:
            The file path of the downloaded video.
        """
        try:
            ytdlp_path = self.get_binary_path('yt-dlp')  # Auto-downloads if needed

            # Ensure output directory exists
            os.makedirs(output_path, exist_ok=True)

            # Run yt-dlp with output template
            output_template = os.path.join(output_path, '%(title)s.%(ext)s')
            result = subprocess.run([
                ytdlp_path, video_url, '-o', output_template
            ], capture_output=True, text=True, check=True)

            # Parse the output to get the actual filename
            lines = result.stdout.split('\n')
            for line in lines:
                if 'has already been downloaded' in line or 'Destination:' in line:
                    # Extract filename from the line
                    filename = line.split()[-1]
                    return filename

            # If we can't parse the exact filename, return the template path
            return output_template

        except subprocess.CalledProcessError as e:
            raise Exception(f"Video download failed: {e.stderr}")

    def download_audio_only(self, video_url: str, output_path: str, audio_format: str) -> str:
        """
        Downloads the audio track from a video and saves it as an audio file.
        
        Args:
            video_url: The URL of the video.
            output_path: The directory to save the audio file in.
            audio_format: The desired audio format (e.g., 'mp3', 'm4a', 'wav').
            
        Returns:
            The file path of the extracted audio.
        """
        try:
            ytdlp_path = self.get_binary_path('yt-dlp')  # Auto-downloads if needed

            # Ensure output directory exists
            os.makedirs(output_path, exist_ok=True)

            # Run yt-dlp with audio extraction
            output_template = os.path.join(output_path, f'%(title)s.{audio_format}')
            result = subprocess.run([
                ytdlp_path, video_url, '-x', '--audio-format', audio_format,
                '-o', output_template
            ], capture_output=True, text=True, check=True)

            # Parse the output to get the actual filename
            lines = result.stdout.split('\n')
            for line in lines:
                if 'has already been downloaded' in line or 'Destination:' in line:
                    filename = line.split()[-1]
                    return filename

            return output_template

        except subprocess.CalledProcessError as e:
            raise Exception(f"Audio download failed: {e.stderr}")

    def download_playlist(self, playlist_url: str, output_path: str) -> str:
        """
        Downloads all videos from a given playlist URL.
        
        Args:
            playlist_url: The URL of the playlist.
            output_path: The directory where the playlist videos will be saved.
            
        Returns:
            The path to the directory containing the downloaded videos.
        """
        try:
            ytdlp_path = self.get_binary_path('yt-dlp')  # Auto-downloads if needed

            # Ensure output directory exists
            os.makedirs(output_path, exist_ok=True)

            # Run yt-dlp for playlist
            output_template = os.path.join(output_path, '%(playlist_index)s - %(title)s.%(ext)s')
            subprocess.run([
                ytdlp_path, playlist_url, '-o', output_template
            ], capture_output=True, text=True, check=True)

            return output_path

        except subprocess.CalledProcessError as e:
            raise Exception(f"Playlist download failed: {e.stderr}")

    def download_video_by_quality(self, video_url: str, output_path: str, quality: str) -> str:
        """
        Downloads a video in a specific quality or resolution.
        
        Args:
            video_url: The URL of the video to download.
            output_path: The directory where the video will be saved.
            quality: The desired quality string (e.g., 'best', '720p', '1080p').
            
        Returns:
            The file path of the downloaded video.
        """
        try:
            ytdlp_path = self.get_binary_path('yt-dlp')  # Auto-downloads if needed

            # Ensure output directory exists
            os.makedirs(output_path, exist_ok=True)

            # Convert quality to yt-dlp format
            if quality == 'best':
                format_selector = 'best'
            elif quality == 'worst':
                format_selector = 'worst'
            elif quality.endswith('p'):
                # For resolution like 720p, 1080p
                height = quality[:-1]
                format_selector = f'best[height<={height}]'
            else:
                format_selector = quality

            output_template = os.path.join(output_path, '%(title)s.%(ext)s')
            result = subprocess.run([
                ytdlp_path, video_url, '-f', format_selector, '-o', output_template
            ], capture_output=True, text=True, check=True)

            # Parse the output to get the actual filename
            lines = result.stdout.split('\n')
            for line in lines:
                if 'has already been downloaded' in line or 'Destination:' in line:
                    filename = line.split()[-1]
                    return filename

            return output_template

        except subprocess.CalledProcessError as e:
            raise Exception(f"Quality-specific video download failed: {e.stderr}")

    def download_subtitles(self, video_url: str, output_path: str, language_code: str) -> str:
        """
        Downloads the subtitles for a video.
        
        Args:
            video_url: The URL of the video.
            output_path: The directory to save the subtitle file in.
            language_code: The language code for the desired subtitles (e.g., 'en', 'es').
            
        Returns:
            The file path of the downloaded subtitle file.
        """
        try:
            ytdlp_path = self.get_binary_path('yt-dlp')  # Auto-downloads if needed

            # Ensure output directory exists
            os.makedirs(output_path, exist_ok=True)

            # Download subtitles only
            output_template = os.path.join(output_path, '%(title)s.%(ext)s')
            result = subprocess.run([
                ytdlp_path, video_url, '--write-subs', '--sub-langs', language_code,
                '--skip-download', '-o', output_template
            ], capture_output=True, text=True, check=True)

            # The subtitle file will have the same name but with .srt extension
            subtitle_file = output_template.replace('.%(ext)s', f'.{language_code}.srt')
            return subtitle_file

        except subprocess.CalledProcessError as e:
            raise Exception(f"Subtitle download failed: {e.stderr}")

    def download_video_with_thumbnail(self, video_url: str, output_path: str) -> str:
        """
        Downloads a video and embeds its thumbnail as cover art.
        
        Args:
            video_url: The URL of the video.
            output_path: The directory where the video will be saved.
            
        Returns:
            The file path of the video with the embedded thumbnail.
        """
        try:
            ytdlp_path = self.get_binary_path('yt-dlp')  # Auto-downloads if needed

            # Ensure output directory exists
            os.makedirs(output_path, exist_ok=True)

            # Download with thumbnail embedding
            output_template = os.path.join(output_path, '%(title)s.%(ext)s')
            result = subprocess.run([
                ytdlp_path, video_url, '--embed-thumbnail', '--write-thumbnail',
                '-o', output_template
            ], capture_output=True, text=True, check=True)

            # Parse the output to get the actual filename
            lines = result.stdout.split('\n')
            for line in lines:
                if 'has already been downloaded' in line or 'Destination:' in line:
                    filename = line.split()[-1]
                    return filename

            return output_template

        except subprocess.CalledProcessError as e:
            raise Exception(f"Video with thumbnail download failed: {e.stderr}")

File: bridges/nodejs/src/sdk/tools/ytdlp-tool.ts
import { execSync } from 'node:child_process'
import { mkdirSync } from 'node:fs'
import { join } from 'node:path'
import { spawn } from 'node:child_process'

import { Tool, type ProgressCallback } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'

export default class YtdlpTool extends Tool {
  private static readonly TOOLKIT = 'video_streaming'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    // Load configuration from central toolkits directory
    // Use class name for tool config name
    const toolConfigName = this.constructor.name
      .toLowerCase()
      .replace('tool', '')
    this.config = ToolkitConfig.load(YtdlpTool.TOOLKIT, toolConfigName)
  }

  get toolName(): string {
    // Dynamic tool name based on class name
    return this.constructor.name
  }

  get toolkit(): string {
    return YtdlpTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Downloads a single video from the provided URL.
   * @param videoUrl The URL of the video to download.
   * @param outputPath The directory where the video will be saved.
   * @returns A promise that resolves with the file path of the downloaded video.
   */
  async downloadVideo(videoUrl: string, outputPath: string): Promise<string> {
    try {
      const ytdlpPath = await this.getBinaryPath('yt-dlp') // Auto-downloads if needed

      // Ensure output directory exists
      mkdirSync(outputPath, { recursive: true })

      // Run yt-dlp with output template
      const outputTemplate = join(outputPath, '%(title)s.%(ext)s')
      const result = execSync(
        `"${ytdlpPath}" "${videoUrl}" -o "${outputTemplate}"`,
        {
          encoding: 'utf8'
        }
      )

      // Parse the output to get the actual filename
      const lines = (result as string).split('\n')
      for (const line of lines) {
        if (
          line.includes('has already been downloaded') ||
          line.includes('Destination:')
        ) {
          const filename = line.split(' ').pop()
          if (filename) return filename
        }
      }

      // If we can't parse the exact filename, return the template path
      return outputTemplate
    } catch (error: unknown) {
      throw new Error(`Video download failed: ${(error as Error).message}`)
    }
  }

  /**
   * Downloads the audio track from a video and saves it as an audio file.
   * @param videoUrl The URL of the video.
   * @param outputPath The directory to save the audio file in.
   * @param audioFormat The desired audio format (e.g., 'mp3', 'm4a', 'wav').
   * @returns A promise that resolves with the file path of the extracted audio.
   */
  async downloadAudioOnly(
    videoUrl: string,
    outputPath: string,
    audioFormat: string
  ): Promise<string> {
    try {
      const ytdlpPath = await this.getBinaryPath('yt-dlp') // Auto-downloads if needed

      // Ensure output directory exists
      mkdirSync(outputPath, { recursive: true })

      // Run yt-dlp with audio extraction
      const outputTemplate = join(outputPath, `%(title)s.${audioFormat}`)
      const result = execSync(
        `"${ytdlpPath}" "${videoUrl}" -x --audio-format ${audioFormat} -o "${outputTemplate}"`,
        {
          encoding: 'utf8'
        }
      )

      // Parse the output to get the actual filename
      const lines = (result as string).split('\n')
      for (const line of lines) {
        if (
          line.includes('has already been downloaded') ||
          line.includes('Destination:')
        ) {
          const filename = line.split(' ').pop()
          if (filename) return filename
        }
      }

      return outputTemplate
    } catch (error: unknown) {
      throw new Error(`Audio download failed: ${(error as Error).message}`)
    }
  }

  /**
   * Downloads all videos from a given playlist URL.
   * @param playlistUrl The URL of the playlist.
   * @param outputPath The directory where the playlist videos will be saved.
   * @returns A promise that resolves with the path to the directory containing the downloaded videos.
   */
  async downloadPlaylist(
    playlistUrl: string,
    outputPath: string
  ): Promise<string> {
    try {
      const ytdlpPath = await this.getBinaryPath('yt-dlp') // Auto-downloads if needed

      // Ensure output directory exists
      mkdirSync(outputPath, { recursive: true })

      // Run yt-dlp for playlist
      const outputTemplate = join(
        outputPath,
        '%(playlist_index)s - %(title)s.%(ext)s'
      )
      execSync(`"${ytdlpPath}" "${playlistUrl}" -o "${outputTemplate}"`, {
        encoding: 'utf8'
      })

      return outputPath
    } catch (error: unknown) {
      throw new Error(`Playlist download failed: ${(error as Error).message}`)
    }
  }

  /**
   * Downloads a video in a specific quality or resolution.
   * @param videoUrl The URL of the video to download.
   * @param outputPath The directory where the video will be saved.
   * @param quality The desired quality string (e.g., 'best', '720p', '1080p').
   * @param onProgress The callback function for progress reporting.
   * @returns A promise that resolves with the file path of the downloaded video.
   */
  async downloadVideoByQuality(
    videoUrl: string,
    outputPath: string,
    quality: string,
    onProgress?: ProgressCallback
  ): Promise<string> {
    try {
      const ytdlpPath = await this.getBinaryPath('yt-dlp') // Auto-downloads if needed

      // Ensure output directory exists
      mkdirSync(outputPath, { recursive: true })

      // Convert quality to yt-dlp format
      let formatSelector: string
      if (quality === 'best') {
        formatSelector = 'best'
      } else if (quality === 'worst') {
        formatSelector = 'worst'
      } else if (quality.endsWith('p')) {
        // For resolution like 720p, 1080p
        const height = quality.slice(0, -1)
        formatSelector = `best[height<=${height}]`
      } else {
        formatSelector = quality
      }

      const outputTemplate = join(outputPath, '%(title)s.%(ext)s')

      return new Promise((resolve, reject) => {
        const childProcess = spawn(ytdlpPath, [
          videoUrl,
          '-f',
          formatSelector,
          '-o',
          outputTemplate,
          '--newline'
        ])

        let downloadedFilePath = outputTemplate

        childProcess.stdout.on('data', (data) => {
          const output = data.toString()
          const lines = output.split('\n')

          for (const line of lines) {
            // Parse download progress
            if (line.includes('[download]')) {
              const progressMatch = line.match(
                /\[download\]\s+(\d+\.?\d*)%\s+of\s+([\d.]+\w+)\s+at\s+([\d.]+\w+\/s)\s+ETA\s+([\d:]+)/
              )
              if (progressMatch && onProgress) {
                onProgress({
                  percentage: parseFloat(progressMatch[1]),
                  size: progressMatch[2],
                  speed: progressMatch[3],
                  eta: progressMatch[4],
                  status: 'downloading'
                })
              }
            }

            // Check for completed download or destination file
            if (
              line.includes('[download] Destination:') ||
              line.includes('has already been downloaded')
            ) {
              const pathMatch =
                line.match(/Destination:\s+(.+)$/) ||
                line.match(/(.+)\s+has already been downloaded/)
              if (pathMatch) {
                downloadedFilePath = pathMatch[1].trim()
              }
            }

            // Check for download completion
            if (line.includes('[download] 100%') && onProgress) {
              onProgress({
                percentage: 100,
                status: 'completed'
              })
            }
          }
        })

        childProcess.stderr.on('data', (data) => {
          const output = data.toString()
          // yt-dlp sometimes outputs progress to stderr as well
          if (output.includes('[download]') && onProgress) {
            const lines = output.split('\n')
            for (const line of lines) {
              const progressMatch = line.match(
                /\[download\]\s+(\d+\.?\d*)%\s+of\s+([\d.]+\w+)\s+at\s+([\d.]+\w+\/s)\s+ETA\s+([\d:]+)/
              )
              if (progressMatch) {
                onProgress({
                  percentage: parseFloat(progressMatch[1]),
                  size: progressMatch[2],
                  speed: progressMatch[3],
                  eta: progressMatch[4],
                  status: 'downloading'
                })
              }
            }
          }
        })

        childProcess.on('close', (code) => {
          if (code === 0) {
            resolve(downloadedFilePath)
          } else {
            reject(
              new Error(
                `Quality-specific video download failed with exit code ${code}`
              )
            )
          }
        })

        childProcess.on('error', (error) => {
          reject(new Error(`Failed to start yt-dlp process: ${error.message}`))
        })
      })
    } catch (error: unknown) {
      throw new Error(
        `Quality-specific video download failed: ${(error as Error).message}`
      )
    }
  }

  /**
   * Downloads the subtitles for a video.
   * @param videoUrl The URL of the video.
   * @param outputPath The directory to save the subtitle file in.
   * @param languageCode The language code for the desired subtitles (e.g., 'en', 'es').
   * @returns A promise that resolves with the file path of the downloaded subtitle file.
   */
  async downloadSubtitles(
    videoUrl: string,
    outputPath: string,
    languageCode: string
  ): Promise<string> {
    try {
      const ytdlpPath = await this.getBinaryPath('yt-dlp') // Auto-downloads if needed

      // Ensure output directory exists
      mkdirSync(outputPath, { recursive: true })

      // Download subtitles only
      const outputTemplate = join(outputPath, '%(title)s.%(ext)s')
      execSync(
        `"${ytdlpPath}" "${videoUrl}" --write-subs --sub-langs ${languageCode} --skip-download -o "${outputTemplate}"`,
        {
          encoding: 'utf8'
        }
      )

      // The subtitle file will have the same name but with .srt extension
      const subtitleFile = outputTemplate.replace(
        '.%(ext)s',
        `.${languageCode}.srt`
      )
      return subtitleFile
    } catch (error: unknown) {
      throw new Error(`Subtitle download failed: ${(error as Error).message}`)
    }
  }

  /**
   * Downloads a video and embeds its thumbnail as cover art.
   * @param videoUrl The URL of the video.
   * @param outputPath The directory where the video will be saved.
   * @returns A promise that resolves with the file path of the video with the embedded thumbnail.
   */
  async downloadVideoWithThumbnail(
    videoUrl: string,
    outputPath: string
  ): Promise<string> {
    try {
      const ytdlpPath = await this.getBinaryPath('yt-dlp') // Auto-downloads if needed

      // Ensure output directory exists
      mkdirSync(outputPath, { recursive: true })

      // Download with thumbnail embedding
      const outputTemplate = join(outputPath, '%(title)s.%(ext)s')
      const result = execSync(
        `"${ytdlpPath}" "${videoUrl}" --embed-thumbnail --write-thumbnail -o "${outputTemplate}"`,
        {
          encoding: 'utf8'
        }
      )

      // Parse the output to get the actual filename
      const lines = (result as string).split('\n')
      for (const line of lines) {
        if (
          line.includes('has already been downloaded') ||
          line.includes('Destination:')
        ) {
          const filename = line.split(' ').pop()
          if (filename) return filename
        }
      }

      return outputTemplate
    } catch (error: unknown) {
      throw new Error(
        `Video with thumbnail download failed: ${(error as Error).message}`
      )
    }
  }
}
